### DISCUSSION TITLE
When using the GCNConv layer, should I add the self loops when building the adjacency matrix?

### DISCUSSION THREAD

**Discussion by zzzy0828** [danielegrattarola/spektral#363] [Q&A] [answered]
Hi! We know that in the paper [Semi-Supervised Classification with Graph Convolutional Networks] the adjacency matrix should be added with self-loops. Now the question is when using the GCNConv layer in Spektral, should I add the self-loops in the adjacency matrix by myself. In other words, should the principal diagonal of the customized adjacency matrix be filled with the value '1'. Or the self-loops have been added in the progress of GCNConv layer's processing, which means the original package/code has included the self loops?
Here's the question I encountered nowadays. Expecting for your reply.
Thank you very much!

**Comment by danielegrattarola** [comment, marked as answer]
Hi, 

you should pre-process the adjacency matrix manually, the layer only computes `A @ X @ W` without changing `A`. 
To do the pre-processing, every layer in Spektral has a `preprocess` method that you can use, so in your case: 

`adj = GCNConv.preprocess(adj)`

More details here: https://graphneural.network/getting-started/

**Reply by zzzy0828** [reply]
Thank you for your reply. Thank you!
